src/JSInterop/Microsoft.JSInterop/src/Infrastructure/DotNetDispatcher.cs

```diff
@@ -381,9 +381,11 @@ private static Task GetTaskByType(Type type, object obj)
 
     private static Task CreateValueTaskConverter<[DynamicallyAccessedMembers(LinkerFlags.JsonSerialized)] T>(object result) => ((ValueTask<T>)result).AsTask();
 
+    [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2111:ReflectionToDynamicallyAccessedMembers",
+        Justification = "The Type being passed into GetOrAdd is annotated, but the trimmer warns anyway. See https://github.com/dotnet/linker/issues/2790")]
     private static (MethodInfo methodInfo, Type[] parameterTypes) GetCachedMethodInfo(IDotNetObjectReference objectReference, string methodIdentifier)
     {
-        var type = objectReference.Value.GetType();
+        var type = objectReference.Type;
```

Comment:
There is a difference here potentially.
```C#
class MyValueBase {}
class MyDerivedValue : MyValueBase {  void ImportantMethod() {} }

void Test()
{
    DotnetObjectReference.Create<MyValueBase>(new MyDerivedValue());
}
```

Before this change the code would build the cache with the `ImportantMethod` in it. After this change the `ImportantMethod` will not be there and thus won't be callable from JS.

We could add an exception/assert into the .ctor of `DotNetObjectReference` that the `value.GetType() == typeof(TValue)`.

Reply:
I think this is an acceptable change, in order to make this path trimmable (which is needed for both wasm and Maui).

> We could add an exception/assert into the .ctor of DotNetObjectReference that the value.GetType() == typeof(TValue).

This would have to be an exception, since the API is public.

Reply:
@SteveSandersonMS do you have an opinion here? Would it be beneficial to add an exception check to `DotNetObjectReference<T>`'s ctor to ensure the Types match?

Reply:
I think this might be problematic as it might prevent some patterns where the type used is a base class (can't remember if we can use interfaces here).

Reply:
Can you point me to them? `DotNetObjectReference<TValue>`'s ctor is internal, and it's only caller is `DotNetObjectReference.Create<TValue>(TValue value)`.

According to https://source.dot.net/#Microsoft.JSInterop/DotNetObjectReference.cs,90b7156b4279e601,references, there are only 3 callers to the Create method, all which don't use base classes.

Reply:
OK, I can change this back to `var type = objectReference.Value.GetType();`. The annotations I'm adding here will allow the "normal" case to work successfully in a trimmed app.

We will still have a trim warning here, which we can add to the baseline. It was previously being suppressed in code, which isn't valid for warnings that are real trimming problems. To address that warning in a future change, maybe we can add a new `DotNetObjectReference.CreateExact` (or something with a better name) that is guaranteed to be safe for trimming. Then the existing `DotNetObjectReference.Create` method can be annotated with `RequiresUnreferencedCode`, so any callers will get trim warnings. And they can switch over to the new, trim-compatible API.

Thoughts on this plan?

Reply:
I've updated the PR to follow the above proposal. Please take a look and let me know what you think.